File: python/tools/insights/create.py
import json

from lib.utils.api import get_project_base_url
from schema.tool_inputs import InsightCreateSchema
from tools.types import Context, TextContent, Tool, ToolResult


async def create_insight_handler(context: Context, params: InsightCreateSchema) -> ToolResult:
    project_id = await context.get_project_id()

    insight_result = await context.api.insights(project_id).create(params.data)

    if not insight_result.success:
        raise Exception(f"Failed to create insight: {insight_result.error}")

    insight_with_url = {
        **insight_result.data.model_dump(),
        "url": f"{get_project_base_url(project_id)}/insights/{insight_result.data.short_id}"
    }

    return ToolResult(content=[TextContent(text=json.dumps(insight_with_url))])


def create_insight_tool() -> Tool[InsightCreateSchema]:
    return Tool(
        name="create-insight",
        description="Creates a new insight in the project.",
        schema=InsightCreateSchema,
        handler=create_insight_handler
    )

File: typescript/src/tools/experiments/create.ts
import { ExperimentCreatePayloadSchema } from "@/schema/experiments";
import { ExperimentCreateSchema } from "@/schema/tool-inputs";
import type { Context, ToolBase } from "@/tools/types";
import type { z } from "zod";

const schema = ExperimentCreateSchema;

type Params = z.infer<typeof schema>;

/**
 * Create a comprehensive A/B test experiment with guided setup
 * This tool helps users create well-configured experiments through conversation
 */
export const createExperimentHandler = async (context: Context, params: Params) => {
	const projectId = await context.stateManager.getProjectId();

	// Transform tool input to API payload format using the schema transformation
	const apiPayload = ExperimentCreatePayloadSchema.parse(params);

	// Send to API with full type safety
	const result = await context.api.experiments({ projectId }).create(apiPayload);

	if (!result.success) {
		throw new Error(`Failed to create experiment: ${result.error.message}`);
	}

	// Format the response with useful information
	const experiment = result.data;
	const experimentWithUrl = {
		...experiment,
		url: `${context.api.getProjectBaseUrl(projectId)}/experiments/${experiment.id}`,
	};

	return {
		content: [
			{
				type: "text",
				text: JSON.stringify(experimentWithUrl, null, 2),
			},
		],
	};
};

const tool = (): ToolBase<typeof schema> => ({
	name: "experiment-create",
	schema,
	handler: createExperimentHandler,
});

export default tool;
